File: commands/furry.ts
import randomcolor from "randomcolor";
import shuffle from "lodash/shuffle";

import { RichEmbed, GuildMember, TextChannel, Message } from "discord.js";
import axios, { AxiosResponse, AxiosRequestConfig } from "axios";

const blacklist: string[] = [
  "cub",
  "children",
  "young",
  "gore",
  "scat",
  "shota",
  "loli",
  "rape",
  "necrophilia",
  "underage",
];

interface Post {
  tags: {
    general: string[];
  };
}

function eGet(
  channel: TextChannel,
  member: GuildMember,
  tags: string[],
  nsfw: boolean,
): void {
  const siteName: string = nsfw ? "e621" : "e926";
  const tags_ = `${tags.join("+")}+-${blacklist.join("+-")}`;
  const url = `https://${siteName}.net/posts.json?limit=100&tags=${tags_}${
    !nsfw ? "+rating:s" : ""
  }`;
  const data: AxiosRequestConfig = {
    method: "get",
    url: url,
    headers: {
      "User-Agent": "RavenBot/1.0 (NotBlizzard)",
    },
  };
  axios(data).then((res: AxiosResponse) => {
    const post = shuffle(
      res.data.posts.filter((post: Post) => {
        return post.tags.general.every((tag: string) => {
          return !blacklist.includes(tag);
        });
      }),
    )[0];

    const embed: RichEmbed = new RichEmbed()
      .setDescription(
        `By **${post.tags?.artist.join(
          ", ",
        )}** | [Link](https://www.${siteName}.net/posts/${post.id})`,
      )
      .addField("Tags Used", tags.join(", "))
      .addField(
        "Votes",
        `:arrow_up: ${post.score.up} | :arrow_down: ${post.score.down}`,
      )
      .setAuthor(member.user.username, member.user.displayAvatarURL)
      .setImage(post.file.url)
      .setColor(randomcolor());
    channel.send(embed);
  });
}

export default {
  e621: {
    args: "e621 <tags>",
    description:
      "search for art on e621 with <tags>. can only be used in NSFW channels.",

    use(
      member: GuildMember,
      channel: TextChannel,
      message: Message,
      args: string[],
    ): Promise<Message | Message[]> | void {
      if (args.filter((tag: string) => blacklist.includes(tag)).length > 0)
        return channel.send("Your search cannot include blacklisted tags.");
      if (!channel.nsfw) {
        channel.send(
          "This channel is not flagged for NSFW. Using e926 instead.",
        );
        return eGet(channel, member, args, false);
      }
      if (args.length === 0) {
        return channel.send(
          `You must have tags. Type "${process.env.RAVEN_COMMAND_KEY}help e621" for more help.`,
        );
      }
      eGet(channel, member, args, true);
    },
  },

  e926: {
    args: "e926 <tags>",
    description: "search for art on e926 with <tags>.",
    use(
      member: GuildMember,
      channel: TextChannel,
      message: Message,
      args: string[],
    ): Promise<Message | Message[]> | void {
      if (args.filter((x) => blacklist.includes(x)).length > 0)
        return channel.send("Your search cannot include blacklisted tags.");
      if (args.length === 0)
        return channel.send(
          `You must have tags. Type "${process.env.RAVEN_COMMAND_KEY}help e926" for more info.`,
        );
      return eGet(channel, member, args, false);
    },
  },

  owo: {
    args: "owo <text>",
    description: "translate <text> to owo language",
    use(
      member: GuildMember,
      channel: TextChannel,
      message: Message,
      args: string[],
    ): Promise<Message | Message[]> {
      const textMessage: string = args
        .join(" ")
        .replace(/(r|l|w)/g, "w")
        .replace(/n(?=a|e|i|o|u)/g, "ny");
      return channel.send(textMessage);
    },
  },
};

File: commands/furry.py

import re
import requests
import random
from discord import Embed, Permissions

blacklist = [
    "cub",
    "children",
    "young",
    "gore",
    "scat",
    "shota",
    "loli",
    "rape",
    "necrophilia",
    "underage",
]

async def get_post(member, channel, tags, nsfw):
    site = "e621" if nsfw else "e926"
    rating = "+rating:s" if not nsfw else ""
    tags_array = tags
    tags = f'{"+".join(tags)}+-{"+-".join(blacklist)}'

    url = f'https://{site}.net/posts.json?limit=100&tags={tags}{rating}'
    headers = {"User-Agent": "RavenBot/1.0 (NotBlizzard)"}
    r = requests.get(url, headers=headers).json()

    posts = [post for post in r["posts"] for tag in post["tags"]["general"] if tag not in blacklist]

    random.shuffle(posts)

    post = posts[0]

    embed = Embed(
        description=f'By {",".join(post["tags"]["artist"])}',
    )

    embed.set_image(url=post["file"]["url"])
    embed.add_field(name="Tags used", value=", ".join(tags_array),inline=False)
    embed.add_field(name="votes",value=f':arrow_up: {post["score"]["up"]} | :arrow_down: {post["score"]["down"]}',inline=False)
    embed.set_author(name=member.name,icon_url=member.avatar_url)
    await channel.send(embed=embed)

async def e621(member, channel, message, args):
    await get_post(member, channel, args, True)

async def e926(member, channel, message, args):
    await get_post(member, channel, args, False)

async def owo(member, channel, message, args):
    message = " ".join(args)
    text = re.sub(r'(r|l|w)', "w", message)
    text = re.sub(r'n(?=a|e|i|o|u)', "ny", text)

    await channel.send(text)